### ISSUE TITLE
Transaction Locks

### ISSUE THREAD

**Issue by advapiIT** [linq2db/linq2db#1910]
I've a quite huge block of code that need to be put under transaction.
This piece of code uses internally other classes that inside opens a new connection against the DB. When the inner connection is open, the query times out
```cs
    using (var db = dataContextFactory.Create())
            {

                using (var transaction = db.BeginTransaction(System.Data.IsolationLevel.ReadCommitted))
                {
SomeCall()
}
}

public void SomeCall()
{
  using (var mainContext = dataContextFactory.Create())
                {
...this locks
}
```

**Comment by sdanyliv** [comment]
You have to reuse connection. Or you have to create them with Transaction parameter - there is constructor for `DataConnection` with accepts transaction.
You are not first one here: #1698

**Comment by advapiIT** [comment]
Thanks for your reply, but is this a leaky abstraction?

**Comment by sdanyliv** [comment]
Why? It is normal behavior. 

- How we should know that you need to reuse transaction for particular connection?
- What to do if application works simultaneously with several databases?

**Comment by advapiIT** [comment]
This works fine in your example since you don't use a Repository

Infact I agree with such a piece of code

public static class UsersData
{
    public static IQueryable<User> GetAvailaibleUsers(IDataContext dc)
    {
          return dc.GetTable<User>().Where(u => u.IsActive);
    }
}

What I (personally) don't like is to expose in the repository the
IDataContext as a parameter (and for now I've passed it as nullable,
so if null I istantiate it internally).

But that's just a thought of mine

Thanks

Paolo

**Comment by sdanyliv** [comment]
- You can create extension methods, so it will looks like native DataContext's methods.
- You can create Service class which accepts DataContext as constructor parameters
- You can extend your repository to accept DataContext as constructor parameter 
- You can setup Dependency Injection to provide the same DataContext to repostories/services.

**Comment by advapiIT** [comment]
Hello @sdanyliv ,
Please excuse me again, but I still have got deadlocks even passing the same connection.

In particular the locks occurs when I do a LoadWith

I explain you what happens in details , hoping you can help me understanding how to fix it.


In the main block of the code I do 

`  var newOpz = db.BenefitRsmPolizzaOpzioni//   dbContext.BENEFIT_RSM_POLIZZA_OPZIONI.AsNoTracking()
                                                .Single(m => m.Id == opzione.Id);

                                            newOpz.IdCondizione = newCat.IdCondizione;
                                            //newOpz.BENEFIT_CONDIZIONI_CATEGORIE = newCat;

                                            newOpz.AltroNomeAX = string.Empty;
                                            newOpz.DataFinePrimoSemestre = null;

                                            // newCat.BENEFIT_RSM_POLIZZA_OPZIONI = new List<BENEFIT_RSM_POLIZZA_OPZIONI>();


                                            var newOpzId = db.InsertWithInt32Identity(newOpz);

                                            newOpz.Id = newOpzId;
`

Then inside a repository where I've passed the connection I need to do

`
 public List<Condizione> GetCondizioniCategoriaByStatusProtocollo(int idPolizza, bool? attiva,
            DateTime dataScadenza, IDataContext mainContext = null)
        {
            try
            {
                bool isDbConnectionPassed = mainContext != null;

                if (!isDbConnectionPassed)
                    mainContext = dataContextFactory.Create();

                try
                {
                  
                    var categorieSP = (from c in ((XXXBENEFITDB)mainContext).BenefitCondizioniCategorie
                            .LoadWith(x => x.BenefitRsmPolizzaOpzionis.FirstOrDefault().BenefitRsmOpzioniCompagnia)
                            .LoadWith(x => x.BenefitCategorie)
                            .LoadWith(x => x.BenefitGruppoRamo)
                            .LoadWith(x => x.BenefitParametri)
                            .LoadWith(x => x.BenefitTipoRischio)

                                       where c.IdPolizza == idPolizza
                        && (attiva == null || (attiva.Value && c.DataAnnullamento == null) ||
                            (!attiva.Value && c.DataAnnullamento != null))
                                       select c).ToList();
`

If I comment the `    .LoadWith(x => x.BenefitRsmPolizzaOpzionis.FirstOrDefault().BenefitRsmOpzioniCompagnia)` it works otherwise not.

I've tried different isolation level but I still have found no solution

What can I do /check?

Thanks

**Comment by sdanyliv** [comment]
I have to review. There can be problem with LoadWith because it may open new connection.

**Comment by advapiIT** [comment]
I'm currently testing this

`   var categorieSP = (from c in ((xxx)mainContext).BenefitCondizioniCategorie
                          //  .LoadWith(x => x.BenefitRsmPolizzaOpzionis.FirstOrDefault().BenefitRsmOpzioniCompagnia)
                            .LoadWith(x => x.BenefitCategorie)
                            .LoadWith(x => x.BenefitGruppoRamo)
                            .LoadWith(x => x.BenefitParametri)
                            .LoadWith(x => x.BenefitTipoRischio)

                                       where c.IdPolizza == idPolizza
                        && (attiva == null || (attiva.Value && c.DataAnnullamento == null) ||
                            (!attiva.Value && c.DataAnnullamento != null))
                                       select c).ToList();


                    foreach(var cat in categorieSP)
                    {
                        cat.BenefitRsmPolizzaOpzionis = ((xxx) mainContext).BenefitRsmPolizzaOpzioni.LoadWith(y=>y.BenefitRsmOpzioniCompagnia) .Where(x=>x.BenefitCondizioniCategorie.IdPolizza == idPolizza && x.IdCondizione == cat.IdCondizione).ToList();
                    }`

and seems to work.. .give you a feedback soon

**Comment by MaceWindu** [comment]
You've hit known issue with current LoadWith implementation #482. We need to create separate connection for each additional LoadWith, wich means each additional query executed in own context, not in context of main query transaction.

I would recommend you to load data using separate queries explicitly for now if you hit issues with current implementation of LoadWith and return back to LoadWith when better eager load implementation #1756 released.

**Comment by MaceWindu** [comment]
Duplicate of #482

**Comment by MaceWindu** [comment]
Fixed by #1756
